### PR TITLE
Devbook: upgrade crates and openapi script

### PR DESCRIPTION
**Pull request by Alenar**
## Content

This PR introduce the `devbook`: a collection of useful scripts and procedures for Mithril developers.

In this `devbook` you will find:

- `upgrade-repository-dependencies`, moved from the runbook since it's more of dev than a run operation
- `upgrade-crates-and-openapi-version`: a new script to easily upgrade all crates versions + the openapi version that are changed in a PR compared to `origin/main`.

## Pre-submit checklist

- Branch
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [x] Update README file (if relevant)

### CODE REVIEW

**Review by sfauvel — APPROVED** [review]

File: docs/devbook/README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Mithril network devbook :shield:
+
+This page gathers the available guides and tools for Mithril nodes developers.
+
+# Guides
+
+| Operation                               | Location                                                                               | Description                                                                         |
+|-----------------------------------------|----------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------|
+| **Upgrade crates and openapi versions** | [upgrade-crates-and-openapi-versions](./upgrade-crates-and-openapi-versions/README.md) | Script to easily upgrade crates and openapi versions before merging a pull requests |
```
Comment (jpraynaud) [inline docs/devbook/README.md:9]:
```suggestion
| **Upgrade crates and openapi versions** | [upgrade-crates-and-openapi-versions](./upgrade-crates-and-openapi-versions/README.md) | Script to easily upgrade crates and openapi versions before merging a pull request |
```

File: docs/devbook/upgrade-crates-and-openapi-versions/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# Upgrade the crates and openapi versions before merging a pull request
+
+## Introduction
+
+This devbook provides a script that allow to automatically upgrade the crates and openapi versions in the project.
```
Comment (jpraynaud) [inline docs/devbook/upgrade-crates-and-openapi-versions/README.md:5]:
```suggestion
This devbook provides a script that allows to automatically upgrade the crates and Open API versions in the project.
```

File: docs/devbook/upgrade-crates-and-openapi-versions/README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+Only crates and openapi with changes on the branch compared to `origin/main` are upgraded.
```
Comment (jpraynaud) [inline docs/devbook/upgrade-crates-and-openapi-versions/README.md:7]:
```suggestion
Only the crates and Open API specifications with changes on the branch compared to `origin/main` are upgraded.
```

File: docs/devbook/upgrade-crates-and-openapi-versions/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+## Prerequisites
+
+It requires having `cargo-get` installed, which can be done with the following command:
```
Comment (jpraynaud) [inline docs/devbook/upgrade-crates-and-openapi-versions/README.md:11]:
```suggestion
It requires to have `cargo-get` installed, which can be done with the following command:
```

File: docs/devbook/upgrade-crates-and-openapi-versions/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+```shell
+. ./docs/devbook/upgrade-crates-and-openapi-versions/upgrade_crates_and_openapi_versions.sh
+```
+
+### Run
+
+Run the script with the `--run` argument to bump the crates and openapi versions.
+
+The script will finish with a preformatted commit message that can be used to create a commit.
```
Comment (jpraynaud) [inline docs/devbook/upgrade-crates-and-openapi-versions/README.md:34]:
```suggestion
The script will output a preformatted commit message that can be used to create a commit when it completes.
```

File: docs/devbook/upgrade-crates-and-openapi-versions/upgrade_crates_and_openapi_versions.sh [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+then
+    echo -e "${ORANGE}warning${RESET}: script is run in dry mode. To execute the changes, run ${GREEN}./update_crate_versions.sh --run${RESET}"
+else
+  UPDATED_CRATES="\n$(find . -name Cargo.toml -exec git diff "$COMMIT_REF" {} + | grep -E "^[\+\-]version = \"[0-9\.]+\"|name = " | tr '\n' ' ' | sed -r "s/ name = \"([a-z\-]+)\" -version = \"([0-9\.]+)\" \+version = \"([0-9\.]+)\" ?/* \1 from \`\2\` to \`\3\`\n/g")"
+  COMMIT_MESSAGE=$(echo -e "chore: upgrade crate versions${OPEN_API_UPDATE_MESSAGE}\n${UPDATED_CRATES}${OPEN_API_UPDATE}")
+
+  echo -e "$COMMIT_MESSAGE"
+
+  # Do we want to commit the changes ?
+  #git add $OPEN_API_FILE Cargo.lock ./*/Cargo.toml ./internal/*/Cargo.toml ./mithril-test-lab/*/Cargo.toml && git commit -m "$COMMIT_MESSAGE"
```
Comment (jpraynaud) [inline docs/devbook/upgrade-crates-and-openapi-versions/upgrade_crates_and_openapi_versions.sh:99]: Maybe this command can also part of the output?

**Review by jpraynaud — APPROVED** [review]
LGTM :+1:

**Review by dlachaume — APPROVED** [review]
LGTM 👍

File: docs/devbook/upgrade-crates-and-openapi-versions/upgrade_crates_and_openapi_versions.sh (hunk 6 above)
Comment (Alenar) [reply docs/devbook/upgrade-crates-and-openapi-versions/upgrade_crates_and_openapi_versions.sh]: I added a `--commit` argument to execute the commit instead.

**Pull request merged**